### ISSUE TITLE
TorchScript Model giving different Results when loaded in Python and Android Sudio

### ISSUE THREAD

**Issue by NaumanHSA** [ultralytics/ultralytics#5117]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
I've been working on running a custom trained segmentation model in Android Studio with NMS. Please review the steps below that I have implemented.

1. Trained a custom YOLOv8n segmentation model, and converted it into torchscript using the export.py.
2. I then loaded the converted model in Python using **torch.jit.load()**. As the exported model doesn't implements NMS, so I tried to export another torchscript model with NMS embedded inside.
3. I prepared all the functions involved in NMS and created another torch **nn.Module** wrapping the preprocessing, exported model predictions and applying NMS. Please see the code snipped below.
4. I then converted the newly created model to torchscript using **optimize_for_mobile(torch.jit.script(...))._save_for_lite_interpreter(...)**.
5. Finally, I loaded the converted model in Android Studio using **pytorch_android_lite** Module.

Now, coming to my actual question. I've trained two same models (YOLOv8n Segmentation) with different datasets, one with 28 classes and another with just one class. Both models have been trained with input image size 640. The problem is that I'm getting desired predictions from one model with 28 classes, but entirely wrong results with the other one. I've tried exactly the same steps as the first one, but it's giving me wrong results. Things I've tried:

1. Loaded the same converted model in Python using **torch.jit.load()** and tested the results. The results were fine.
2. Tried to convert the model with and without **optimize_for_mobile()**.
3. Tried to convert the model with **torch.jit.script()** and **torch.jit.trace()**. Nothing works.
4. Tried to export the YOLOv8 model with and without the optimize parameter.
5. Tried to retrain the model with different input image size.
6. Tried every possible debugging to check if there's something wrong with the NMS code. But it works with the first model just fine.

![YOLOv8n_torchscript_wrapped_model](https://github.com/ultralytics/ultralytics/assets/22982600/1a9a07d0-6599-4bf8-a76b-b8668e423047)

The results my model is returning a list of two tensors; bounding boxes and masks.
`List[Tensor(num_detections * [x1, y1, x2, y2, conf_score, cls]), Tensor(num_detections * [input_size, input_size])]`

Please, if someone have already solved this issue, assist me. Thanks in Advance !!!

**Comment by glenn-jocher** [comment]
@NaumanHSA hello,

Thank you for detailing your question. It indeed is strange that you're observing varying results from two models with different class sizes, while both being deployed in the same way.

From what you have described, it seems the deployment process is managed appropriately and limits the possibilities of bugs during conversion to TorchScript and application of NMS.

From the information provided, it seems like the model with a single class is the one giving incorrect results. Since you've trained the models under identical parameters with only the datasets varying, the problem could potentially lie within the single-class dataset itself. It might be worth looking at if the annotations and divisions (train/val/test splits) in this dataset have been done correctly and are consistent with the 28-class dataset.

Also, since you've already experimented with different image input sizes during retraining, it might be beneficial to experiment with other hyperparameters such as the learning rate, weight decay, or number of iterations to see if it improves model performance on the single-class dataset.

Remember to ensure the preprocessing of your inputs in your Android deployment is consistent with what has been done during your model training. Small differences in aspects like normalization, image resizing or data augmentation could lead to very different model predictions. 

Lastly, if you haven't done so already, closely inspect the outputs you're receiving from the single class model on Android. There might be a pattern or specific type of error occurring which could help diagnose the issue. 

Please keep in mind that model deployment involves many moving parts, and differences between runtime environments (like Python and Android) can sometimes lead to subtle, unforeseen discrepancies. 

Best of luck with your debugging and please do not hesitate to reach out if you require additional assistance!

**Comment by NaumanHSA** [comment]
@glenn-jocher,
Thank you so much for the quick response, and the details provided, that's what I love about you guys. I'm gonna put my comments on above mentioned points.

**Training**
Since I've prepared both datasets following the exact same steps, I don't think there's a problem with the dataset. I used LabelMe for annotations. Then converted JSON files to YOLO format, and split into train/test using distributed sampling, all using a single Python script. Yet I've not trained with different parameters. But my question is, does retraining the model really matters since I'm getting the desired results in Python, i) with original YOLO model and ii) my converted Torchscript model with NMS?

**Preprocessing**
As it was a bit difficult for me to implement all steps of preprocessing in Android, so I embedded a few steps in the torchscript model. It receives a Tensor of shape (3, imgsz, imgsz) and do the other preprocessing. Please see the code snipped below. However, on the Android side, I've implemented resizing the image (Mat) using OpenCV inspired by YOLO **LetterBox** class. That I'm sure is working perfectly fine. Once resized, I'm simply converting it into a Tensor.

![YOLOv8n_torchscript_wrapped_model_preprocessing](https://github.com/ultralytics/ultralytics/assets/22982600/875eebe3-c99b-48df-b257-02e12a36e1c3)

**Inspecting Models Outputs**
After not getting the desired results, I removed the NMS part from my model and inspected the outputs. From my inspection, the confidence scores that model returns are very low, that's why filtering out detections based on **conf_threshold** returns empty tensor, and eventually that leads into returning an empty results. Lowering the **conf_threshold** to 0.1 gave some wrong results.


At this point, I don't know what should I try, but surely I'll retrain the model with higher input size, different hyperparameters, and may I need to add an extra class that I won't use (it may be because of single class segmentation?). Thank you.

**Comment by NaumanHSA** [comment]
@glenn-jocher,
I think I have found the issue, and have resolved it.
So I tried to do random changes in the code and analyze the results. Finally, I found something that became the cause for wrong outputs. While converting an OpenCV Mat object to Pytorch tensor on the Android side, I first got the data from Mat object into bytes array, and then created a tensor from bytes using Tensor.fromBlob(). Then a marical happened when I changed the bytes array to float array. I still don't know the exact reason, but it worked in my case. Now, I'm getting the desired outputs from both models. See the code snipped below for Mat to Tensor conversion.

![Java_Mat2Tensor](https://github.com/ultralytics/ultralytics/assets/22982600/96081c0f-0a1b-488d-a2d2-9c37a0007fe1)


I think you we now mark the issue as resolved.

**Comment by glenn-jocher** [comment]
@NaumanHSA, that's great to hear that you've found a solution! Sometimes, it's the intricacies like data type during conversion that can make a significant difference - it emphasizes the importance of ensuring every step aligns accurately. 

Converting the OpenCV Mat object to a float array rather than a byte array before transforming it into a Pytorch tensor seems to have worked for you. This implies a potential mismatch of data type expectations between OpenCV and PyTorch, leading to incorrect outputs.

Thank you for sharing the resolution, as this could be beneficial for other users experiencing similar issues. Your efforts to debug and resolve this matter are very much appreciated! We're glad you've achieved the desired results from your models now.

I'll proceed to mark this issue as resolved. Don't hesitate to reach out if you face any other issues in the future. Happy coding!
